File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    
    static int maxDistance;
    
    public int solution(int n, int[][] edge) {
        int answer = 0;
        int[] distance = new int[n+1];
        List<Integer>[] list = new ArrayList[n+1];
        boolean[] visited = new boolean[n+1];
        maxDistance = 0;
        
        for(int i = 1; i <= n; i++){
            list[i] = new ArrayList<>();
        }
        
        for(int[] w : edge){
            list[w[0]].add(w[1]);
            list[w[1]].add(w[0]);
        }
        
        bfs(visited, list, distance);

        for(int i = 1; i <= n; i++){
            if(distance[i] == maxDistance) answer++;
        }
        
        return answer;
    }
    
    static void bfs(boolean[] visited, List<Integer>[] list,  int[] distance){
        
        Deque<int[]> queue = new ArrayDeque<>();
        queue.add(new int[] {1, 0});
        visited[1] = true;
        
        while(!queue.isEmpty()){
            
            int[] curr = queue.poll();
            int v = curr[0];
            int d = curr[1];
            
            for(int w : list[v]){
                
                if(!visited[w]){
                    visited[w] = true;
                    queue.add(new int[] {w, d+1});
                    
                    if(distance[w] != 0){
                        distance[w] = Math.min(distance[w], d + 1);
                    }else{
                         distance[w] = d + 1;
                    }
                    
                    maxDistance = Math.max(maxDistance, distance[w]);
                }
            }
            
        }
        
    }
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    li = [[] for _ in range(n + 1)]
    for a, b in edge:
        li[a].append(b)
        li[b].append(a)

    dist = [-1] * (n + 1)
    dist[1] = 0

    q = deque([1])
    
    while q:
        v = q.popleft()
        for nv in li[v]:
            if dist[nv] == -1: 
                dist[nv] = dist[v] + 1
                q.append(nv)

    maxd = max(dist[1:])
    return dist.count(maxd)
